### ISSUE TITLE
Orientation from EXIF

### ISSUE THREAD

**Issue by jameswood** [wildlyinaccurate/jekyll-responsive-image#46]
Most cameras encode JPGs to be "right-side-up" when the camera is held normally. If the camera is rotated, the JPG is encoded the same way, but with an EXIF flag to indicate how the viewing software or device should rotate the image for correct orientation.

From what I can tell (and I really am sorry if I've messed something up!) jekyll-responsive-image does not take this into account, and produces resized images of the "uncorrected" orientation.

Is there a way to have jekyll-responsive-image respect the rotation flag from the exif file and produce corrected (rotated) image sets? This would avoid the need to manually process these images.

Rotating the image in CSS with `image-orientation: from-image;` only partially solves the problem, because then the width and height dimensions are rightly swapped.

Correct orientation :
![original image](https://cloud.githubusercontent.com/assets/1140049/22002781/adb14332-dca2-11e6-96ff-7e54433e53dc.png)

Resized image as it appears on site:
![resized image](https://cloud.githubusercontent.com/assets/1140049/22001606/689f9422-dc9a-11e6-8570-4432c0db44a2.png)

Thanks for any tips!
J

**Comment by jameswood** [comment]
It appears that [ImageMagick can do this](http://imagemagick.org/Usage/photos/#orient) with the `-auto-orient` operator. I don't know enough about Ruby to know whether or not rmagick supports this feature, or where in the gem I could add it... but I'm having a go at sherlocking my way out of trouble so I thought I'd mention it here in case it helps!

**Comment by wildlyinaccurate** [comment]
Hey @jameswood, sorry for the slow reply on this! I'll set some time aside this week to make sure that this plugin respects the orientation of the source image.

**Comment by jameswood** [comment]
@wildlyinaccurate please... no need to apologise at all. Thanks (on behalf of everyone who uses it...) for any attention you can spare!

**Comment by wildlyinaccurate** [comment]
I wrote a test (#47) asserting that the EXIF orientation is preserved, expecting it to fail... But it passes, because it turns out the EXIF metadata is not lost when images are resized. So I'm guessing that I misunderstood what you meant.

It seems like what you _actually_ want is for jekyll-responsive-image to perform a rotation transformation based on the EXIF orientation when it resizes an image. Is that correct? If that's the case, I can make that an extra configuration option - something like `auto_orient: true`. I'm wary to make it the default behaviour because there will undoubtedly be website owners who will get a surprise when their images are suddenly oriented differently.

But yeah, if that's what you're after, I can get that done pretty quickly. You already found the right rmagick method, I'll just need to write some tests to go with it :grin:

**Comment by jameswood** [comment]
Hey Joseph,

Yes, I did mean that it would be helpful for me if the plugin could correct the orientation of generated images based on their `exif-rotation` tag, because otherwise some photos will scale and display incorrectly (or need to be manually rotated/re-saved). I was not referring to keeping the exif metadata intact through the process (but I imagine that keeping/stripping exif metadata could be a really useful option to have for some users!).

I totally agree that you shouldn't introduce problems/surprises for existing users. But, as far as I can imagine, the only images this change would break are those with incorrect `exif rotation` tags which are currently being inadvertently corrected by this bug. These images will currently be resizing and displaying differently than they would in most image editing/viewing software. So, my feeling is that you'd be pretty safe in enabling this change by default, given the assumption that any affected images currently in the wild have likely been manually edited to avoid this issue already, or they're presently being displayed with incorrect orientation that no one has noticed. Or, my logic is bad!

To put it another way, I think the "expected behaviour" would be to correct the rotation by default, but you've obviously thought about this for much longer and more deeply than I have. (Sometimes an outside perspective can be helpful - I hope I'm not sounding pushy!)

Default or not, I'm looking forward to the update. Thanks again for your time!

**Comment by jameswood** [comment]
Is there anything I can do to help with this?

**Comment by wildlyinaccurate** [comment]
Hey, sorry I've gone quiet. I've been pretty busy with work & life at the moment and haven't had any time to focus on my open source stuff.

In terms of what you can do to help, there was one thing I hadn't thought through fully:

We'll want to rotate the original image as well as the resized ones. Do we leave the source image untouched and only rotate the one that is copied to the _site folder? Or do we try to be "helpful" and rotate the source one too? I'd lean towards the first option, because I don't want to ruin the quality of the source image.

You're also more than welcome to take a crack at implementing it. Rotating the _resized_ images should be simple, and could be done in `ResizeHandler#resize_image`. But this plugin currently doesn't touch the original image, so there'd need to be a new code path that rotates the source image.

**Comment by jameswood** [comment]
Please! No apology needed! I was worried that I'd overstepped or something!

I don't know the first thing about ruby, or I'd help. It is time for me to learn the first thing about ruby, and then help. I'll give it a go! Thanks for the pointer! Wish me luck...

**Comment by jameswood** [comment]
[I gave it a go!](https://github.com/wildlyinaccurate/jekyll-responsive-image/pull/54)

I added a preference (disabled by default) to rotate images before rescaling them. I had some trouble where it worked fine here but died on my Digital Ocean VPS, because RMagick was using all the ram. Turns out `img.auto_orient!` was the way to go. What a ride! Sorry for spamming with more commits and pull requests than were strictly necessary...

How weirdly awesome is Ruby's backwards IF?! `img.auto_orient! if config['respect_exif_rotation']` - I love it!

**Comment by wildlyinaccurate** [comment]
Fixed in #55, and released as v1.2.0.

Thanks again for your help, @jameswood. The biggest barrier for me dedicating time to my open source projects is always motivation. Your contribution is hugely appreciated ⭐️

**Comment by jameswood** [comment]
Thanks for your help, encouragement and inspiration... as well as your hard work on a genuinely helpful tool!
